src/net/peer/connect/connection_info.rs

```diff
@@ -24,10 +28,12 @@ pub struct PrivConnectionInfo<UID> {
     pub id: UID,
     #[doc(hidden)]
     pub for_direct: Vec<SocketAddr>,
-    #[doc(hidden)]
-    pub for_hole_punch: Vec<SocketAddr>,
-    #[doc(hidden)]
-    pub hole_punch_socket: Option<TcpBuilder>,
+    // P2P connection info is optional in case when p2p `rendezvous_connect` fails.
+    // In such cases we can still proceed with direct connections, but we must keep track
+    // that P2P connections should not be attempted.
+    pub p2p_conn_info: Option<Bytes>,
+    pub rendezvous_channel: UnboundedBiChannel<Bytes>,
+    pub connection_rx: Option<oneshot::Receiver<Result<TcpStream, SingleConnectionError>>>,
```

Comment:
It looks like `p2p_conn_info` and `connection_rx` shouldn't be two separate `Option`s. If they're always both `Some` or both `None` then they should be combined into a single `Option`. Create a new type to put in the `Option` if necessary.